spring-data-microservice-transaction-sample/README.md

```diff
@@ -0,0 +1,548 @@
+...
+$ ./gradlew :client:run --args="PlaceOrder 1 3:1,4:1"
+...
+{"order_id": "b6adabd8-0a05-4109-9618-3420fea3161f"}
+...
+```
+
+## Clean up
+
+To stop Cassandra, MySQL and the Microservices, run the following command:
+
+```shell
+$ docker-compose down
+```
+
+## How the microservice transaction is achieved
+
+So far, you have run the sample application, but you haven't seen how it is implemented.
+Let's look at the code to see how the transaction that spans the services is implemented.
+
+Transaction #2 (Placing an order) achieves the microservice transaction, so we focus on this transaction here.
+
+The sequence diagram of transaction #2 is as follows:
+
+![Sequence Diagram](images/sequence_diagram.png)
+
+### 1. Start a two-phase commit transaction
+
+When a client sends `Place an order` request to Order Service, [OrderService.placeOrder()](order-service/src/main/java/sample/order/OrderService.java#L97) is called, and the microservice transaction starts.
+
+At first, Order Service starts a two-phase commit transaction by calling the repository class's [executeTwoPcTransaction()](order-service/src/main/java/sample/order/OrderService.java#L100-L102):
+
+```java
+execAndReturnResponse(responseObserver, "Placing an order", () -> {
+  // Start a two-phase commit transaction
+  TwoPcResult<String> result = orderRepository.executeTwoPcTransaction(txId -> {
+```
+
+The following `Execute CRUD operations`, `Two-phase Commit` and `Error handling` are automatically performed inside the API.
+
+### 2. Execute CRUD operations
+
+After the transaction is started, the CRUD operations are executed by `executeTwoPcTransaction()`.
+
+Order Service puts the order information to the `order_service.orders` table also the detailed information to `order_service.statements` (the code is [here](order-service/src/main/java/sample/order/OrderService.java#L106-L128)):
+
+```java
+// Put the order info into the orders table
+orderRepository.insert(order);
+
+AtomicInteger amount = new AtomicInteger();
+for (ItemOrder itemOrder : request.getItemOrderList()) {
+  int itemId = itemOrder.getItemId();
+  int count = itemOrder.getCount();
+  // Retrieve the item info from the items table
+  Optional<Item> itemOpt = itemRepository.findById(itemId);
+  if (!itemOpt.isPresent()) {
+    String message = "Item not found: " + itemId;
+    responseObserver.onError(
+        Status.NOT_FOUND.withDescription(message).asRuntimeException());
+    throw new ScalarDbNonTransientException(message);
+  }
+  Item item = itemOpt.get();
+
+  int cost = item.price * count;
+  // Put the order statement into the statements table
+  statementRepository.insert(new Statement(itemId, orderId, count));
+  // Calculate the total amount
+  amount.addAndGet(cost);
+}
+```
+
+And, Order Service calls the `payment` gRPC endpoint of Customer Service along with the transaction ID (the code is [here](order-service/src/main/java/sample/order/OrderService.java#L131)).
```

Comment:
```suggestion
And, the Order Service calls the `payment` gRPC endpoint of the Customer Service along with the transaction ID (the code is [here](order-service/src/main/java/sample/order/OrderService.java#L131)).
```

Reply:
Applying this change in https://github.com/scalar-labs/scalardb-samples/pull/40